libcxx/src/debugging.cpp

```diff
@@ -0,0 +1,201 @@
+// running under the debugger or has a debugger attached post facto).
+static bool __is_debugger_present() noexcept {
+  // Technical Q&A QA1361: Detecting the Debugger
+  // https://developer.apple.com/library/archive/qa/qa1361/_index.html
+
+  // Initialize mib, which tells 'sysctl' to fetch the information about the current process.
+
+  array mib{CTL_KERN, KERN_PROC, KERN_PROC_PID, ::getpid()};
+
+  // Initialize the flags so that, if 'sysctl' fails for some bizarre
+  // reason, we get a predictable result.
+
+  struct kinfo_proc info {};
+
+  // Call sysctl.
+  // https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/sysctl.3.html
+
+  size_t info_size = sizeof(info);
+  if (::sysctl(mib.data(), mib.size(), &info, &info_size, nullptr, 0) != 0) {
+    _LIBCPP_ASSERT_INTERNAL(false, "'sysctl' runtime error");
+    return false;
+  }
+
+  // If the process is being debugged if the 'P_TRACED' flag is set.
+  // https://github.com/freebsd/freebsd-src/blob/7f3184ba797452703904d33377dada5f0f8eae96/sys/sys/proc.h#L822
+
+#    if defined(__FreeBSD__)
+  const auto p_flag = info.ki_flag;
+#    else // __APPLE__
+  const auto p_flag = info.kp_proc.p_flag;
+#    endif
+
+  return ((p_flag & P_TRACED) != 0);
+}
+
+#  elif defined(__linux__)
+
+static bool __is_debugger_present() noexcept {
+#    if defined(_LIBCPP_HAS_NO_FILESYSTEM)
+  _LIBCPP_ASSERT_INTERNAL(false,
+                          "Function is not available. Could not open '/proc/self/status' for reading, libc++ was "
+                          "compiled with _LIBCPP_HAS_NO_FILESYSTEM.");
+  return false;
+#    elif defined(_LIBCPP_HAS_NO_LOCALIZATION)
+  _LIBCPP_ASSERT_INTERNAL(false,
+                          "Function is not available. Could not open '/proc/self/status' for reading, libc++ was "
+                          "compiled with _LIBCPP_HAS_NO_LOCALIZATION.");
+  return false;
+#    else
+  // https://docs.kernel.org/filesystems/proc.html
+
+  // Get the status information of a process by reading the file /proc/PID/status.
+  // The link 'self' points to the process reading the file system.
+  ifstream status_file{"/proc/self/status"};
+  if (!status_file.is_open()) {
+    _LIBCPP_ASSERT_INTERNAL(false, "Could not open '/proc/self/status' for reading.");
+    return false;
+  }
+
+  std::string token;
+  while (status_file >> token) {
+    // If the process is being debugged "TracerPid"'s value is non-zero.
+    if (token == "TracerPid:") {
+      int pid;
+      status_file >> pid;
+      return pid != 0;
+    }
+    getline(status_file, token);
+  }
+
+  return false;
+#    endif // _LIBCPP_HAS_NO_FILESYSTEM
+}
+
+#  elif defined(_AIX)
+
+_LIBCPP_HIDE_FROM_ABI bool __is_debugger_present() noexcept {
+  // Get the status information of a process by memory mapping the file /proc/PID/status.
+  // https://www.ibm.com/docs/en/aix/7.3?topic=files-proc-file
+  char filename[] = "/proc/4294967295/status";
```

Comment:
```suggestion
char filename[] = "/proc/4294967295/status";
  char filename[] = "/proc/XXXXXXXXXX/status";
```

Does this value hold any significance? Can we do this instead?